Mathlib/Data/ListM/Basic.lean

```diff
@@ -180,6 +221,22 @@ unsafe def filterMapM [Alternative m] {α β : Type u} (f : α → m β) : ListM
       (f a >>= fun b => return (some b, filterMapM f r)) <|> return (none, filterMapM f r)
 #align tactic.mllist.mfilter_map ListM.filterMapM
 
+/-- Take the initial segment of the lazy list, until the function `f` first fails. -/
+unsafe def takeWhileM [Alternative m] (f : α → m β) : ListM m α → ListM m α
```

Comment:
I feel like checking for success is a bit of a Lean 3-ism, an API like `takeWhileM (f : α → m (ULift Bool)) ...` is more natural now and also consistent with `List.filterM`.

Reply:
Okay. I liked that Lean 3-ism, but I agree it is not necessarily the obvious semantics!

Reply:
Hmm, I really like the `filterMapM` for being able to discard cases where the monadic function fails. I feel like every time I call this it is going to be with a failing tactic that with this change I need to `try ... catch ...`.

What about a primed version of these, that assume `[Alternative m]` and have the old behaviour?

Reply:
> I feel like every time I call this it is going to be with a failing tactic that with this change I need to try ... catch ....

We have a `succeeds` function for that, and `xs.takeWhileM (succeeds do checkType ·)` looks pretty nice and explicit.

But I don't want to block this PR on that change.  Most functions on `ListM` have this implicit succeeds-API.  I know you that you need the PR for more interesting code. :smile:

Reply:
Oh and for `filterMapM` there's `try?` to convert failure to Option.

Reply:
Okay, I'm convinced, and have changed the API.